### DISCUSSION TITLE
Description of -fast option

### DISCUSSION THREAD

**Discussion by ebraun68** [iqtree/iqtree2#240] [Q&A] [answered]
I was curious about the -fast option. The only description I can find in the manual is this:

Turn on the fast tree search mode, where IQ-TREE will just
construct two starting trees: maximum parsimony and BIONJ, which
are then optimized by nearest neighbor interchange (NNI).
Introduced in version 1.6.

I have a few questions about the details of this search:

1) Does IQ-TREE do NNIs on both starting trees, or does it check the likelihood of the MP and BIONJ trees and then choose one of those trees for further branch swapping?

2) How stochastic is this feature given different runs (or different random number seeds)? Is stochasticity coming in for the choice of MP starting tree, in the order of the NNIs, or both?

3) Can a -fast search be combined with a user starting tree?

I apologize if I missed any of this when I tried to find it in the manual

**Comment by bqminh** [comment, marked as answer]
Thanks for asking, we should have documented it more clearly. This example command:

```
iqtree2 -s example.phy -m GTR+G -fast
```
will do the following steps (with comparison to a normal search when necessary):

1. Load the alignment and skip sequence composition test of a normal search. 
2. Construct a parsimony tree with stepwise addition + SPR search and initialise candidate tree set with this one tree.
3. Estimate model parameters (GTR+G in this case) on the parsimony tree with an optimisation epsilon 0.5 (normal: 0.1).
4. Compute pairwise distances based on the model and a NJ tree from the distance matrix.
5. If NJ tree is different from parsimony tree, put it into the candidate tree set.
6. For each tree in the candidate set (which has 1 or 2 trees): 
  6.1. Optimize it from NNI
  6.2. If it has higher likelihood than the best tree, reestimate model parameters with epsilon 0.5 (normal: 0.1) and add it to the candidate set.
7. Reestimate model parameters on the best tree in the candidate set with epsilon 0.05 (normal: 0.01).
8. Report the best tree and model parameters as output of IQ-TREE.

Now your questions:

Q1. NNI is done both both starting trees, according to this workflow.
Q2. Stochasticity is only happening in step 2.
Q3. If you supply a starting tree via `-t TREE_FILE` option: Step 2 is changed to using this tree and steps 4-5 won't do the NJ tree to keep the candidate set to have this one tree. The remaining steps stay the same.

**Reply by ebraun68** [reply]
Thanks so much for this answer. I would love to really understand the settings and the timing values in the .log file, so I did a few experiments, which motivated a few additional questions. Anyway, I tried -t TREE_FILE, using trees generated by parsimonator so I would have a record of the starting tree topologies and then went over the results. Here are my new questions:

Q1. Is there a way to save the MP tree? I know you normally use 100 MP starting trees, so I assume that is why you avoided writing the MP trees unlike the .bionj tree

Q2. Is the epsilon = 0.5 in steps 3 and 6.2 (if the NJ tree is better than MP) simply the value of --epsilon (if you pass the value). i.e., if I used --epsilon 0.1 would the epsilon for steps 3 and 6.2 be 1 and the value for step 7 be 0.1

Q3. When optimizing parameters in step 3 (and 6.2) is most of the time spent optimizing the GTR matrix parameters and gamma shape parameters or the branch lengths? I assume the latter.

Q4. Same as Q3, but for step 7. Also, do the branch lengths in the .treefile change after optimization? 

Q5. In the .log file, are the values after each step CPU time or wall-clock time? For example, in one test run I got the following:

This was for a dataset with 2760 sequences, 895617 columns, 794189 distinct patterns,
365126 parsimony-informative, 130187 singleton sites, 400304 constant sites

Estimate model parameters (epsilon = 0.500)
1. Initial log-likelihood: -29402364.645
2. Current log-likelihood: -27217404.568
3. Current log-likelihood: -27182411.900
4. Current log-likelihood: -27178888.315
5. Current log-likelihood: -27178463.330
6. Current log-likelihood: -27178404.538
7. Current log-likelihood: -27178393.806
8. Current log-likelihood: -27178391.008
9. Current log-likelihood: -27178389.926
Optimal log-likelihood: -27178389.531
Rate parameters:  A-C: 1.01242  A-G: 4.92573  A-T: 0.56001  C-G: 1.75996  C-T: 4.89427  G-T: 1.00000
Base frequencies:  A: 0.299  C: 0.200  G: 0.199  T: 0.301
Gamma shape alpha: 0.343
Parameters optimization took 9 rounds (9616.553 sec)
Wrote distance file to... 

I assume this means 9616 seconds of CPU time was used here. 

I hope I'm not getting too far into the weeds with this. I really appreciate the time you spent answering the earlier question!
